NIFI-5406: Adding new listing strategy to List processors.

ijokarumawak opened the pull request:
Although it has many TODO items, I thought it's a good idea to share the new listing model early so that everyone can comment on it. Thanks.

----


Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [ ] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

markap14 commented on nifi-nar-bundles/nifi-extension-utils/nifi-processor-utils/src/main/java/org/apache/nifi/processor/util/list/ListedEntityTracker.java:
```diff
@@ -0,0 +1,322 @@
+    private String getCacheKey() {
+        switch (scope) {
+            case LOCAL:
+                return format("%s::%s::%s", getClass().getSimpleName(), componentId, nodeId);
```
We should avoid using `getClass().getSimpleName()` here. If ever this class were to be renamed, it would reset all users' state for all instances. Instead, we should use a hardcoded value of "ListedEntityTracker".

ijokarumawak replied:
Done, I used 'ListedEntities' as prefix.

markap14 commented on nifi-nar-bundles/nifi-extension-utils/nifi-processor-utils/src/main/java/org/apache/nifi/processor/util/list/ListedEntityTracker.java:
```diff
@@ -0,0 +1,322 @@
+        oldEntityIds.forEach(oldEntityId -> alreadyListedEntities.remove(oldEntityId));
+
+        // Emit updated entities.
+        updatedEntities.forEach(updatedEntity -> {
```
This again feels like a spot where imperative coding style is more natural than lambdas, given that it's a simple for-each with a significantly sized 'body'. The lambda here doesn't really buy anything but leads to much more cryptic stack traces and is more expensive to evaluate.

ijokarumawak replied:
Done.

markap14 commented on nifi-nar-bundles/nifi-extension-utils/nifi-processor-utils/src/main/java/org/apache/nifi/processor/util/list/ListedEntityTracker.java:
```diff
@@ -0,0 +1,322 @@
+            logger.debug("Removed old entities count: {}, Updated entities count: {}",
+                    new Object[]{oldEntityIds.size(), updatedEntities.size()});
+            if (logger.isTraceEnabled()) {
+                logger.trace("\"Removed old entities: {}, Updated entities: {}",
```
Was this log message intended to begin with a double-quote?

ijokarumawak replied:
Thanks, it was copy-pasting. Removed the double-quote.

markap14 commented on nifi-nar-bundles/nifi-extension-utils/nifi-processor-utils/src/main/java/org/apache/nifi/processor/util/list/ListedEntityTracker.java:
```diff
@@ -0,0 +1,322 @@
+
+public class ListedEntityTracker<T extends ListableEntity> {
+
+    private ObjectMapper objectMapper = new ObjectMapper();
```
Should probably mark member variable as `final` if not changing.

ijokarumawak replied:
Done.

markap14 commented on nifi-nar-bundles/nifi-extension-utils/nifi-processor-utils/src/main/java/org/apache/nifi/processor/util/list/ListedEntityTracker.java:
```diff
@@ -0,0 +1,322 @@
+public class ListedEntityTracker<T extends ListableEntity> {
+
+    private ObjectMapper objectMapper = new ObjectMapper();
+    private volatile Map<String, ListedEntity> alreadyListedEntities;
```
This is a member variable that is mutable. Access to it is not protected within the code, which means that it is not thread-safe. Probably best to use a ConcurrentHashMap.

ijokarumawak replied:
Updated to use ConcurrentHashMap. Thanks.

markap14 commented on nifi-nar-bundles/nifi-extension-utils/nifi-processor-utils/src/main/java/org/apache/nifi/processor/util/list/ListedEntity.java:
```diff
@@ -0,0 +1,44 @@
+     */
+    private long size;
+
+    public void setTimestamp(long timestamp) {
```
I don't think these are ever called other than immediately after creating the object, to set the initial value. If they are, then this is not thread-safe but is accessed by multiple threads. If they are not, then it probably makes sense to just provide the size and timestamp as constructor arguments, no? At a minimum, it makes it clear when looking at the code that it's immutable and therefore threadsafe.

ijokarumawak replied:
You are right. I was using final fields and constructor, but when it's deserialized, I got an Exception around Jackson, so I added setters. I've added Jackson annotations to use custom constructor for deserialization. Now it's clear those are immutable.

markap14 commented on nifi-nar-bundles/nifi-extension-utils/nifi-processor-utils/src/main/java/org/apache/nifi/processor/util/list/ListedEntityTracker.java:
```diff
@@ -0,0 +1,322 @@
+
+    private Serializer<String> stringSerializer = (v, o) -> o.write(v.getBytes(StandardCharsets.UTF_8));
+    private Serializer<Map<String, ListedEntity>> listedEntitiesSerializer
+            = (v, o) -> objectMapper.writeValue(new GZIPOutputStream(o), v);
```
When writing to a GZIPOutputStream, closing the stream is very important because it writes a trailer to the OutputStream. As far as I can tell, this stream is not being closed, so this is not really a proper GZIP stream. We should ensure that close the GZIPOutputStream. Also, by default, GZIPOutputStream uses a compression level of 5. We've done a huge amount of testing GZIPOutputStream against text and all results show that a compression level of 1 instead of 5 provides almost the same compression level (say, it gets us 98% of the way there) but compression level 1 is much less expensive. I'd recommend using a compression level of 1 as a result. Easiest way to do this is to use the `org.apache.nifi.stream.io.GZIPOutputStream`, which extends `java.util.zip.GZIPOutputStream` but allows you to provide a compression level.

ijokarumawak replied:
Thanks for catching this. I switched to use NiFi's GZIPOutputStream. As for closing stream, I don't think we can close the output stream passed to the serializer, because the stream may be used by the caller to write additional bytes. None of current implementation does that, but we can't be certain about if it's fine to close the stream here in this context.

GZIPOutputStream has `finish` method to finish writing Gzip data without closing the underlying stream. So I added calling finish from serializer.

ijokarumawak commented on nifi-nar-bundles/nifi-extension-utils/nifi-processor-utils/src/main/java/org/apache/nifi/processor/util/list/AbstractListProcessor.java:
```diff
@@ -169,6 +179,28 @@
+
+    public static final AllowableValue BY_ENTITIES = new AllowableValue("entities", "Tracking Entities",
+            "This strategy tracks information of all the listed entities within the latest 'Entity Tracking Time Window' to determine new/updated entities." +
+                    " See 'Entity Tracking Time Window' description for detail on how it works." +
```
I wanted to guide user to look at 'Entity Tracking Time Window' property's description. Updated the text.

ijokarumawak commented on nifi-nar-bundles/nifi-extension-utils/nifi-processor-utils/src/main/java/org/apache/nifi/processor/util/list/ListedEntityTracker.java:
```diff
@@ -0,0 +1,322 @@
+    }
+
+    static void validateProperties(ValidationContext context, Collection<ValidationResult> results, Scope scope) {
+        final Consumer<PropertyDescriptor> validateRequiredProperty = property -> {
```
I just wanted to avoid adding another private method just for this validateProperties method. But I agree with you, private method would increase readability. Updated.

ijokarumawak commented on nifi-nar-bundles/nifi-extension-utils/nifi-processor-utils/src/main/java/org/apache/nifi/processor/util/list/ListedEntityTracker.java:
```diff
@@ -0,0 +1,322 @@
+        validateRequiredProperty.accept(ListedEntityTracker.TRACKING_TIME_WINDOW);
+
+        if (Scope.LOCAL.equals(scope)) {
+            if (StringUtils.isEmpty(context.getProperty(NODE_IDENTIFIER).evaluateAttributeExpressions().getValue())) {
```
Done.

ijokarumawak commented on nifi-nar-bundles/nifi-extension-utils/nifi-processor-utils/src/main/java/org/apache/nifi/processor/util/list/ListedEntityTracker.java:
```diff
@@ -0,0 +1,322 @@
+        }
+    }
+
+    static void setCurrentTimestampSupplier(Supplier<Long> supplier) {
```
Yes, the function is used to use consistent system timestamp from unit tests. And yes, it can be fragile under multi-threaded unit testing if we use `static` one here. I've switched to use final instance field passed by constructor. Thanks.

ijokarumawak commented:
@markap14 I believe all review comments are addressed now. Also, fixed few things found during tests under clustered environment. And changed default initial listing target from 'window' to 'all' to make it behave the same as existing listing does. I hope it's ready to be merged. Thanks for reviewing and insightful feedback!

markap14 commented:
@ijokarumawak thanks for the updates! Code looks good! I will run a few tests and if all looks good will merge to master. Thanks!

markap14 commented:
All of my testing looks good as well. +1 merged to master. Thanks for the improvement, @ijokarumawak !
